lib/src/component/pure_ui_component.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+
+/// A mixin to make a [UiComponent2] instance behave
+/// like a [ReactJS `PureComponent`](https://reactjs.org/docs/react-api.html#reactpurecomponent).
+///
+/// If your component does not support `props.children`, you can override [supportsPropChildren]
+/// to `false` for further optimizations.
+mixin PureUiComponent<T extends UiProps> on UiComponent2<T> {
+  @mustCallSuper
+  @override
+  bool shouldComponentUpdate(Map nextProps, Map nextState) {
+    final currentPropsWithoutChildren = Map.of(this.props)..remove('children');
+    final nextPropsWithoutChildren = Map.of(nextProps)..remove('children');
+
+    bool shouldUpdate = true;
+
+    try {
+      shouldUpdate = !_deepCollectionEquality.equals(currentPropsWithoutChildren, nextPropsWithoutChildren) ||
```

Comment:
We shouldn't be doing deep equality by default, since that's more expensive and not [what `PureCopmonent` does](https://reactjs.org/docs/react-api.html#reactpurecomponent). And `PureComponent` actually does identity equality, which we should be doing too.

Also, `MapEquality` isn't optimal, so we should use a lighter weight function.

Here's a lighter implementation we can use instead, based on [`shallowEqual` from Redux JS](https://github.com/reduxjs/react-redux/blob/master/src/utils/shallowEqual.js)
```dart

/// Returns whether maps [a] and [b] have an
/// have [identical] sets of values for the same keys.
//
// Ported from https://github.com/reduxjs/react-redux/blob/573db0bfc8d1d50fdb6e2a98bd8a7d4675fecf11/src/utils/shallowEqual.js
//
// The MIT License (MIT)
//
// Copyright (c) 2015-present Dan Abramov
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the 
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish, 
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
bool shallowIdentical(Map a, Map b) {
  if (identical(a, b)) return true;
  if (a.length != b.length) return false;
  for (final key in a.keys) {
    if (!b.containsKey(key)) return false;
    if (identical(b[key], a[key])) return false;
  }
  return true
}

```

Reply:
I think I'm going to add that utility to my changes in #462, so we'll just have to make sure there's only one copy based on who merges first

Reply:
Can you be more specific what you mean by identity equality?

Reply:
Sorry, using `identical`(https://api.dart.dev/stable/2.8.4/dart-core/identical.html) as opposed to using the `==` operator

Reply:
So the implementation should just be:
```dart
bool shouldComponentUpdate(Map nextProps, Map nextState) =>
    areMapsShallowIdentical(props, nextProps) && areMapsShallowIdentical(state, nextState);
```

Reply:
Doesn't seem like it will be this easy.  If we use `identical` on prop values that are functions - the component redraws every time there is a function as a prop value, even if the function isn't an anonymous closure...

Reply:
If I update `areShallowMapsIdentical` like so - everything works as expected, but I'm not sure how you feel about this.  Seems like a bug that the same named function from the same object passed as a value within a map is not `identical`... but I'm not 100% sure about that.

```dart
bool areMapsShallowIdentical(Map a, Map b) {
  if (identical(a, b)) return true;
  if (a.length != b.length) return false;
  for (final key in a.keys) {
    if (!b.containsKey(key)) return false;
    final bVal = b[key];
    final aVal = a[key];
    if (bVal is Function && aVal is Function) {
      if (bVal != aVal) return false;
    }
    if (!identical(bVal, aVal)) return false;
  }
  return true;
}
```

Reply:
Yeah - I think its that the JS interop anonymizes those closures within `allowInterop` maybe?

Reply:
> Seems like a bug that the same named function from the same object passed as a value within a map is not identical

But they are equal? Do you have an example?

Reply:
@greglittlefield-wf https://github.com/Workiva/over_react/pull/499/files#diff-a9e1386c1bc08edefc73e84b817c5cc0R38-R48

Reply:
And yes - doing a simple `==` on the values returns `true`, but `identical` returns false.

Reply:
Reduced test case adapted from https://dart.dev/guides/language/language-tour#testing-functions-for-equality: https://dartpad.dev/3c2776f4c9d0ab04f238365f84362ba3

Reply:
Resolved in https://github.com/Workiva/over_react/pull/499/commits/eacedbb717ea6550cdc094d4b6bdd05304354792